sycl-fusion/passes/target/TargetFusionInfo.cpp

```diff
@@ -0,0 +1,267 @@
+  bool InsertWrappers =
+      ((StartWrapperFunc && !StartWrapperFunc->isDeclaration()) &&
+       (FinishWrapperFunc && !FinishWrapperFunc->isDeclaration()));
+  auto *WrapperFuncTy = FunctionType::get(
+      Type::getVoidTy(LLVMMod->getContext()), /*isVarArg*/ false);
+  for (auto &BB : *KernelFunc) {
+    for (auto Inst = BB.begin(); Inst != BB.end();) {
+      if (auto *CB = dyn_cast<CallBase>(Inst)) {
+        if (CB->getCalledFunction()->getName().starts_with("__itt_offload")) {
+          Inst = Inst->eraseFromParent();
+          continue;
+        }
+      }
+      ++Inst;
+    }
+    if (InsertWrappers) {
+      if (ReturnInst *RI = dyn_cast<ReturnInst>(BB.getTerminator())) {
+        auto *WrapperCall =
+            CallInst::Create(WrapperFuncTy, FinishWrapperFunc, "", RI);
+        WrapperCall->setCallingConv(CallingConv::SPIR_FUNC);
+      }
+    }
+  }
+  if (InsertWrappers) {
+    KernelFunc->getEntryBlock().getFirstInsertionPt();
+    auto *WrapperCall =
+        CallInst::Create(WrapperFuncTy, StartWrapperFunc, "",
+                         &*KernelFunc->getEntryBlock().getFirstInsertionPt());
+    WrapperCall->setCallingConv(CallingConv::SPIR_FUNC);
+  }
+}
+
+void SPIRVTargetFusionInfo::createBarrierCall(IRBuilderBase &Builder,
+                                              int BarrierFlags) const {
+  if (BarrierFlags == -1) {
+    return;
+  }
+  assert((BarrierFlags == 1 || BarrierFlags == 2 || BarrierFlags == 3) &&
+         "Invalid barrier flags");
+
+  static const auto FnAttrs = AttributeSet::get(
+      LLVMMod->getContext(),
+      {Attribute::get(LLVMMod->getContext(), Attribute::AttrKind::Convergent),
+       Attribute::get(LLVMMod->getContext(), Attribute::AttrKind::NoUnwind)});
+
+  static constexpr StringLiteral N{"_Z22__spirv_ControlBarrierjjj"};
+
+  Function *F = LLVMMod->getFunction(N);
+  if (!F) {
+    constexpr auto Linkage = GlobalValue::LinkageTypes::ExternalLinkage;
+
+    auto *Ty = FunctionType::get(
+        Builder.getVoidTy(),
+        {Builder.getInt32Ty(), Builder.getInt32Ty(), Builder.getInt32Ty()},
+        false /* isVarArg*/);
+
+    F = Function::Create(Ty, Linkage, N, *LLVMMod);
+
+    F->setAttributes(
+        AttributeList::get(LLVMMod->getContext(), FnAttrs, {}, {}));
+    F->setCallingConv(CallingConv::SPIR_FUNC);
+  }
+
+  // See
+  // https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Memory_Semantics_-id-
+  SmallVector<Value *> Args{
+      Builder.getInt32(/*Exec Scope : Workgroup = */ 2),
+      Builder.getInt32(/*Exec Scope : Workgroup = */ 2),
+      Builder.getInt32(0x10 | (BarrierFlags % 2 == 1 ? 0x100 : 0x0) |
+                       ((BarrierFlags >> 1 == 1 ? 0x200 : 0x0)))};
+
+  auto *BarrierCallInst = Builder.CreateCall(F, Args);
+  BarrierCallInst->setAttributes(
+      AttributeList::get(LLVMMod->getContext(), FnAttrs, {}, {}));
+  BarrierCallInst->setCallingConv(CallingConv::SPIR_FUNC);
```

Comment:
We have [createSPIRVCall](https://github.com/intel/llvm/blob/2dcc58171668695e93f7f2ef5bbda70cf994c1e9/sycl-fusion/passes/kernel-fusion/Builtins.cpp#LL512C22-L512C30) with this construct. Feel free to rebase to use and refactor as needed.